tools/state-viewer/src/latest_witnesses.rs

```diff
@@ -38,22 +66,86 @@ impl LatestWitnessesCmd {
-                println!("{:#?}", witness);
-            } else if self.binary {
-                println!("{:?}", borsh::to_vec(witness).unwrap());
-            } else {
-                println!("{:?}", witness);
+            match self.mode {
+                LatestWitnessesMode::Pretty => {
+                    println!("{:#?}", witness);
+                    println!("");
+                }
+                LatestWitnessesMode::Binary { ref output_dir } => {
+                    let file_name = format!(
+                        "witness_{}_{}_{}.bin",
+                        witness.chunk_header.height_created(),
+                        witness.chunk_header.shard_id(),
+                        witness.epoch_id.0
+                    );
+                    let file_path = output_dir.join(file_name);
+                    std::fs::write(&file_path, borsh::to_vec(witness).unwrap()).unwrap();
+                    println!("Saved to {:?}", file_path);
+                }
             }
-            println!("");
         }
     }
 }
+
+#[derive(clap::Parser)]
+pub struct ValidateWitnessCmd {
+    /// File with state witness saved as raw &[u8].
+    #[arg(long)]
+    input_file: PathBuf,
+}
+
+impl ValidateWitnessCmd {
+    pub(crate) fn run(&self, home_dir: &Path, near_config: NearConfig, store: Store) {
+        let encoded_witness: Vec<u8> =
+            std::fs::read(&self.input_file).expect("Failed to read file");
+        let witness: ChunkStateWitness = borsh::BorshDeserialize::try_from_slice(&encoded_witness)
+            .expect("Failed to deserialize witness");
+        let chain_genesis = ChainGenesis::new(&near_config.genesis.config);
+        let epoch_manager =
+            EpochManager::new_arc_handle(store.clone(), &near_config.genesis.config);
+        let runtime_adapter =
+            NightshadeRuntime::from_config(home_dir, store, &near_config, epoch_manager.clone())
+                .expect("could not create the transaction runtime");
+        let shard_tracker = ShardTracker::new(
+            TrackedConfig::from_config(&near_config.client_config),
+            epoch_manager.clone(),
+        );
+        // TODO(stateless_validation): consider using `ChainStore` instead of
+        // `Chain`.
+        let chain = Chain::new_for_view_client(
+            Clock::real(),
+            epoch_manager.clone(),
+            shard_tracker,
+            runtime_adapter.clone(),
+            &chain_genesis,
+            DoomslugThresholdMode::TwoThirds,
+            false,
+        )
+        .unwrap();
+        let processing_done_tracker = ProcessingDoneTracker::new();
+        let waiter = processing_done_tracker.make_waiter();
+        chain
+            .shadow_validate_state_witness(
+                witness,
+                epoch_manager.as_ref(),
+                runtime_adapter.as_ref(),
+                Some(processing_done_tracker),
+            )
+            .unwrap();
+        waiter.wait();
```

Comment:
:heart:

A message like "OK" would be nice, just to make it clear that the validation passed.

Reply:
Hmm, the waiter doesn't return anything by default... I think `ProcessingDoneTracker/Waiter` must become a `OnceCell<bool>` to propagate result properly, otherwise it can't be retrieved.
Can I make it a separate issue and ask caller to use `RUST_LOG=debug` for now? :)

Reply:
Oh you're right, for a moment I thought that it either fails at `unwrap()` or succeeds, but it doesn't work like that.